src/python/dxpy/utils/config.py

```diff
@@ -0,0 +1,297 @@
+        except ValueError as e:
+            warn("WARNING: Expected _DX_DEBUG to be an integer, but got", environ["_DX_DEBUG"])
+            dxpy._DEBUG = 0
+
+        self._user_conf_dir = expanduser(environ.get("DX_USER_CONF_DIR", "~/.dnanexus_config"))
+
+        dxpy._UPGRADE_NOTIFY = os.path.join(self._user_conf_dir, ".upgrade_notify")
+        # If last upgrade notification was less than 24 hours ago, disable it
+        if os.path.exists(dxpy._UPGRADE_NOTIFY) and os.path.getmtime(dxpy._UPGRADE_NOTIFY) > time.time() - 86400:
+            dxpy._UPGRADE_NOTIFY = False
+
+        env_vars = self._read_conf_dir(self.get_global_conf_dir())
+        env_vars.update(self._read_conf_dir(self.get_user_conf_dir()))
+        env_vars.update(self._read_conf_dir(self.get_session_conf_dir()))
+        env_overrides = []
+        for var in self.VAR_NAMES:
+            if var in environ:
+                if var in env_vars and env_vars.get(var) != environ[var]:
+                    env_overrides.append(var)
+                env_vars[var] = environ[var]
+
+        for var in env_vars:
+            if env_vars[var] is not None:
+                environ[var] = env_vars[var]
+
+        if sys.stdout.isatty():
+            if not suppress_warning and len(env_overrides) > 0:
+                msg = "WARNING: The following environment variables were found to be different than the values " + \
+                      "last stored by dx: "
+                warn(fill(msg + ", ".join(env_overrides), width=80))
+                msg = "To use the values stored by dx, unset the environment variables in your shell by running " + \
+                      '"source ~/.dnanexus_config/unsetenv".  To clear the dx-stored values, run "dx clearenv".'
+                warn(fill(msg, width=80))
+
+        self._sync_dxpy_state()
+
+    def _sync_dxpy_state(self):
+        dxpy.set_api_server_info(host=environ.get("DX_APISERVER_HOST", None),
+                                 port=environ.get("DX_APISERVER_PORT", None),
+                                 protocol=environ.get("DX_APISERVER_PROTOCOL", None))
+
+        if "DX_SECURITY_CONTEXT" in environ:
+            dxpy.set_security_context(json.loads(environ["DX_SECURITY_CONTEXT"]))
+
+        if "DX_JOB_ID" in environ:
+            dxpy.set_job_id(environ["DX_JOB_ID"])
+            if "DX_WORKSPACE_ID" in environ:
+                dxpy.set_workspace_id(environ["DX_WORKSPACE_ID"])
+            if "DX_PROJECT_CONTEXT_ID" in environ:
+                dxpy.set_project_context(environ["DX_PROJECT_CONTEXT_ID"])
+        else:
+            if "DX_PROJECT_CONTEXT_ID" in environ:
+                dxpy.set_workspace_id(environ["DX_PROJECT_CONTEXT_ID"])
+                dxpy.set_project_context(environ["DX_PROJECT_CONTEXT_ID"])
+
+    def get_global_conf_dir(self):
+        return self._global_conf_dir
+
+    def get_user_conf_dir(self):
+        return self._user_conf_dir
+
+    def get_session_conf_dir(self, cleanup=True):
```

Comment:
Perhaps `cleanup` should default to False and only be set to True when initializing. It seems superfluous to attempt the cleanup when saving and clearing the config, too.